File: year/2018/10/solve.jl
using DelimitedFiles
using Plots

function readPts(filename)
	f = open(filename)
	lines = readlines(f)
	close(f)

	data = zeros(Int64, length(lines), 4)
	for i = 1:length(lines)
		line = lines[i]
		line = split(line, ',')
		x = parse(Int, split(line[1], '<')[2])
		yvx = split(line[2], '>')
		y = parse(Int, yvx[1])
		vx = parse(Int, split(yvx[2], '<')[2])
		vy = parse(Int, line[3][1:end-1])
		data[i,1] = x
		data[i,2] = y
		data[i,3] = vx
		data[i,4] = vy
		println(x, " ", y, " ", vx, " ", vy)
	end
	return data
end


function drawPoints(pts, i)
	# mirror y coordinates and plot
	scatter(pts[:,1], pts[:,2].-2*(pts[:,2]), title=string(i))
end


function updatePts(pts)
	for i = 1:size(pts)[1]
		pts[i,1] += pts[i,3]
		pts[i,2] += pts[i,4]
	end
	return pts
end


function fireworks(pts)
	i = 0
	while true
		display(drawPoints(pts, i))
		sleep(0.001)
		pts = updatePts(pts)
		i+=1
	end
end

# part 1 and 2
pts = readPts("day10.input")
fireworks(pts)

File: year/2024/01/solve.py
with open("input.txt", 'r') as f:
    lines = f.readlines()

lx, rx = [], []
for line in lines: # O(n)
    l, r = map(int, line.strip().split())
    lx.append(l)
    rx.append(r)

# O(n log n)
lx = sorted(lx)
rx = sorted(rx)

# *
dist = 0
for i, l in enumerate(lx): # O(n)
    dist += abs(l - rx[i])
print(dist)

# **
from collections import Counter
counter = Counter(rx)

dist = 0
for l in lx: # O(n)
    if l in counter:
        dist += l*counter[l]
print(dist)